ml-agents/mlagents/trainers/settings.py

```diff
@@ -312,6 +344,151 @@ def apply(self, key: str, env_channel: EnvironmentParametersChannel) -> None:
+            if len(reward_buffer) < 1:
+                return False, smoothing
+            measure = np.mean(reward_buffer)
+            if math.isnan(measure):
+                return False, smoothing
+            if increment_condition.signal_smoothing:
+                measure = 0.25 * smoothing + 0.75 * measure
+                smoothing = measure
+            if measure > increment_condition.threshold:
+                return True, smoothing
+        return False, smoothing
+
+
+@attr.s(auto_attribs=True)
+class Lesson:
+    """
+    Gathers the data of one lesson for one environment parameter including its name,
+    the condition that must be fullfiled for the lesson to be completed and a sampler
+    for the environment parameter. If the completion_criteria is None, then this is
+    the last lesson in the curriculum.
+    """
+
+    value: ParameterRandomizationSettings
+    name: str
+    completion_criteria: Optional[CompletionCriteriaSettings] = attr.ib(default=None)
+
+
+@attr.s(auto_attribs=True)
+class EnvironmentParameterSettings:
+    """
+    EnvironmentParameterSettings is an ordered list of lessons for one environment
+    parameter.
+    """
+
+    curriculum: List[Lesson]
+
+    @staticmethod
+    def _check_lesson_chain(lessons, parameter_name):
+        """
+        Ensures that when using curriculum, all non-terminal lessons have a valid
+        CompletionCriteria
+        """
+        num_lessons = len(lessons)
+        for index, lesson in enumerate(lessons):
+            if index < num_lessons - 1 and lesson.completion_criteria is None:
+                raise TrainerConfigError(
+                    f"A non-terminal lesson does not have a completion_criteria for {parameter_name}."
+                )
+
+    @staticmethod
+    def structure(d: Mapping, t: type) -> Dict[str, "EnvironmentParameterSettings"]:
+        """
+        Helper method to structure a Dict of EnvironmentParameterSettings class. Meant
+        to be registered with cattr.register_structure_hook() and called with
+        cattr.structure().
+        """
+        if not isinstance(d, Mapping):
+            raise TrainerConfigError(
+                f"Unsupported parameter environment parameter settings {d}."
+            )
+        d_final: Dict[str, EnvironmentParameterSettings] = {}
+        for environment_parameter, environment_parameter_config in d.items():
+            if (
+                isinstance(environment_parameter_config, Mapping)
+                and "curriculum" in environment_parameter_config
+            ):
+                d_final[environment_parameter] = strict_to_cls(
+                    environment_parameter_config, EnvironmentParameterSettings
+                )
+                EnvironmentParameterSettings._check_lesson_chain(
+                    d_final[environment_parameter].curriculum, environment_parameter
+                )
+            else:
```

Comment:
So, everything under environment_parameters gets converted into a `Lesson`.  My concern with this is that this sets the precedent that every new env parameter controller needs to go through the control for curriculum.

Reply:
I actually think this is OK. A new sampler type (like adaptive, or active learning, etc.) would be set as a Value in the 1st lesson. Running it through curriculum would allow us, for instance, to do multiple steps (e.g. Lesson 0 is a constant value, Lesson 1 is a regular Sampler, Lesson 2 is an adaptive sampler).

Reply:
So we are basically saying the curriculum is the base class for all future environment parameter controllers?

Reply:
We can always modify environment parameter as needed but sampling was easy to fold as curriculum. I think for environment parameters, whatever method we find for generating the parameters, it should be possible to fold it into a curriculum, even if it has only one lesson.

Reply:
My feeling is (1) this unnecessarily incurs the cost of being a curriculum (i.e. the extra function calls in the EnvParMan) when it doesn't need to be and (2) environment parameter controllers may use update rules that are not as rigid as curriculum update thresholds or depend on something other metric than progress/reward.

Reply:
(1) The cost of being a curriculum allows us to have a much simpler interface of the EnvironmentParameterManager. I agree this is a (very small) added cost of computation but reusing code is important IMO. 
(2) Nothing prevents us to make the completion criteria evolve in the future to fit a future use case or create adaptive samplers. We can also add new fields and methods to the EnvironmentParameterManager later if we need, but for this PR, formulating the samplers as one lesson curriculum seemed like an opportunity for code simplification (since curriculum now support sampling).
We can try to edit the design doc with an alternative to this implementation?
Would the EnvironmentParameterSetting have an optional list of lessons and an optional sampler instead of a list of lessons ?